src/kanata.rs

```diff
@@ -253,20 +310,78 @@ impl Kanata {
         Ok(())
     }
 
-    fn check_handle_layer_change(&mut self) {
+    fn check_handle_layer_change(&mut self, tx: &Sender<EventNotification>) {
         let cur_layer = self.layout.current_layer();
         if cur_layer != self.prev_layer {
+            let new = self.layer_info[cur_layer].name.clone();
             self.prev_layer = cur_layer;
             self.print_layer(cur_layer);
+
+            match tx.try_send(EventNotification::LayerChange { new }) {
+                Ok(_) => {}
+                Err(error) => {
+                    log::error!("could not sent event notification: {}", error);
+                }
+            }
         }
     }
 
     fn print_layer(&self, layer: usize) {
-        log::info!("Entered layer:\n{}", self.layer_strings[layer]);
+        log::info!("Entered layer:\n{}", self.layer_info[layer].cfg_text);
+    }
+
+    pub fn start_notification_loop(kanata: Arc<Mutex<Self>>, rx: Receiver<EventNotification>) {
```

Comment:
May need to think about TCP timeout for the clients, e.g. have a heartbeat event sent every 30s (the processing loop can keep track of the timer for this one).

Also handling (and ignoring) the RX on the TCP socket so that the kernel buffers don't fill up.